File: 28. Find the Index of the First Occurrence in a String/find_the_index_of_the_first_occurence_in_a_string.java
class find_the_index_of_the_first_occurence_in_a_string{
	public static boolean CASE_SENSITIVE = false;
	
	public static boolean compare_string (String text1, String text2){
		boolean out = false;
		int len_text1 = text1.length();
		int len_text2 = text2.length();
		
		if ((len_text1 == 0) && (len_text2 == 0)){
			out = true;
			return (out);
		}
		
		if (len_text1 != len_text2){
			return (out);
		}
		
		String text1_c = text1;
		String text2_c = text2;
		
		if (CASE_SENSITIVE){
			text1_c = text1.toUpperCase();
			text2_c = text2.toUpperCase();
		}
		
		for (int i = 0; i < len_text1; i++){
			if (i == 0){
				out = (text1_c.charAt(i) == text2_c.charAt(i));
			} else {
				out = out && (text1_c.charAt(i) == text2_c.charAt(i));
			}
		}
		
		return (out);
	}
	
	public static int first_occurence (String needle, String haystack){
		int len_needle = needle.length();
		int len_haystack = haystack.length();
		int nje = -1;
		
		if (len_needle > len_haystack){
			nje = -1;
		} else if (len_needle == len_haystack){
			nje = (compare_string(needle, haystack)) ? 0 : -1;
		} else {
			int len_searchable = len_haystack - len_needle + 1;
			
			for (int i = 0; i < len_searchable; i++){
				String substring = haystack.substring(i, i + len_needle);
				
				if (compare_string(needle, substring)){
					nje = i;
					break;
				}
			}
		}
		
		return (nje);
	}
	
	public static void main (String[] args){
		String haystack = "leetcode";
		String needle = "ode";
		
		System.out.println("This is the handy work of Java.\n\n\n");
		System.out.println(first_occurence(needle, haystack));
		System.out.println("\n\n\n");
	}
}

File: 28. Find the Index of the First Occurrence in a String/find_the_index_of_the_first_occurence_in_a_string.kt
val CASE_SENSITIVE: Boolean = false;

fun compare_string (text1: String, text2: String) : Boolean{
	var out: Boolean = false;
	val len_text1 = text1.length;
	val len_text2 = text2.length;
	
	if ((len_text1 == 0) && (len_text2 == 0)){
		out = true;
		return (out);
	}
	
	if (len_text1 != len_text2){
		return (out);
	}
	
	var text1_c: String = text1;
	var text2_c: String = text2;
	
	if (CASE_SENSITIVE){
		text1_c = text1.toUpperCase();
		text2_c = text2.toUpperCase();
	}
	
	out = (text1_c == text2_c);
	
	return (out);
}

fun first_occurence (needle: String, haystack: String) : Int{
	val len_needle: Int = needle.length;
	val len_haystack: Int = haystack.length;
	
	var nje: Int = -1;
	
	if (len_needle > len_haystack){
		nje = -1;
	} else if (len_needle == len_haystack){
		nje = if (compare_string(needle, haystack)) 0 else -1;
	} else {
		val len_searchable: Int = len_haystack - len_needle + 1;
		
		for (i in 0 until len_searchable){
			var substring = haystack.substring(i , i + len_needle);
			
			if (compare_string(needle, substring)){
				nje = i;
				break;
			}
		}
	}
	
	return (nje);
}

fun main () {
	val haystack: String = "leetcode";
	val needle: String = "et";
	
	println("This is the handy work of Kotlin.\n\n\n");
	println(first_occurence(needle, haystack));
	println("\n\n\n");
}
